boa/src/syntax/lexer/comment.rs

```diff
@@ -0,0 +1,66 @@
+use super::{Cursor, Error, Tokenizer};
+use crate::syntax::ast::{Position, Span};
+use crate::syntax::lexer::{Token, TokenKind};
+use std::io::Read;
+
+pub(super) struct SingleLineComment;
+
+/// Lexes a single line comment.
+///
+/// Assumes that the initial '//' is already consumed.
+///
+/// More information:
+///  - [ECMAScript reference][spec]
+///  - [MDN documentation][mdn]
+///
+/// [spec]: https://tc39.es/ecma262/#sec-comments
+/// [mdn]: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Lexical_grammar
+impl<R> Tokenizer<R> for SingleLineComment {
+    fn lex(&mut self, cursor: &mut Cursor<R>, start_pos: Position) -> Result<Token, Error>
+    where
+        R: Read,
+    {
+        // Skip either to the end of the line or to the end of the input
+        while let Some(ch) = cursor.next()? {
+            if ch == '\n' {
+                break;
+            }
+        }
+        Ok(Token::new(
+            TokenKind::Comment,
+            Span::new(start_pos, cursor.pos()),
+        ))
+    }
+}
```

Comment:
Checking at the spec:
> Because a single-line comment can contain any Unicode code point except a LineTerminator code point, and because of the general rule that a token is always as long as possible, a single-line comment always consists of all code points from the // marker to the end of the line. However, the LineTerminator at the end of the line is not considered to be part of the single-line comment; it is recognized separately by the lexical grammar and becomes part of the stream of input elements for the syntactic grammar. This point is very important, because it implies that the presence or absence of single-line comments does not affect the process of automatic semicolon insertion (see 11.9).

I don't think we should emit a `Comment` token kind, but I don't think we should emit a `LineTerminator` either, since it's not part of the comment.

The only place where this is relevant is if we have a JS file where its last line is a single line comment, and doesn't have a `\n` in the end. We should just discard the comment, and there should be no `LineTerminator` being produced (since this would affect automatic semicolon insertion).

So, to make things simpler, we can make this ignore the last `\n` by only peeking it, and leaving the lexer produce a `LineTerminator` after this one, and just return the comment.
The issue with the current implementation is that it will remove the line terminator from the lexer list of tokens.

This:
```javascript
let a = 10 // Comment
let b = 20;
```

Will be equivalent to this:
```javascript
let a = 10 let b = 20;
```
But it should be equivalent to this:
```javascript
let a = 10
let b = 20;
```

Note that the correct representation will correctly create two variables, since the line break between the `a` and `b` declaration will create a semicolon automatically. In the current implementation, this will just fail, since we cannot have a `let` in another `let` declaration.

I would add this as a test case.

Reply:
> About benchmarks, It seems that the parser is now **2x-3x** slower. Except, as expected, for long files, where we get a huge speed bump (from 6.1±0.25ms to 742.7±54.50ns), which is an **87.8%** speed improvement.
> 
> I will review the parser, especially the parser cursor, to see if further improvements can be made.

It would be interesting to see a benchmark with a lot of goal symbol switches vs one without

Reply:
Test case added + modified lexer so single line comments no longer consume the last line terminator.